### PR TITLE
integration: wait for ReadyNotify in Issue3699 test

### LINKED ISSUE #5147: TestIssue3699 fails
```
=== RUN   TestIssue3699
--- FAIL: TestIssue3699 (7.64s)
    cluster_test.go:336: unexpected error on Set (client: etcd cluster is unavailable or misconfigured)
```

https://semaphoreci.com/coreos/etcd/branches/pull-request-5145/builds/4

### PR DESCRIPTION
**Pull request by heyitsanthony**
Fixes #5147

### CODE REVIEW

**Comment by gyuho** [comment]
LGTM. thanks.

File: integration/cluster_test.go [hunk 1]
```diff
@@ -319,9 +320,12 @@ func TestIssue3699(t *testing.T) {
 	// node a will remain useless as long as d is the leader.
 	err := c.Members[0].Restart(t)
 	select {
+	// 5 seconds seems to be too brief
```
Comment (xiang90) [inline integration/cluster_test.go]: this comment is confusing... probably just comment about why do we want to wait for 10 seconds instead?

**Comment by xiang90** [comment]
The integration tests failed with leaking routines. I restarted a few times, but still not working. Probably we need to investigate into it before merging this pr.

**Comment by heyitsanthony** [comment]
@xiang90 the rafthttp transport stops before the peer http server stops. A new connection goes through the http server which creates a new pipeline via AddRemote. Since the transport is already stopped it won't be stopped again and so the pipeline goroutines leak.

I added a `panic` to `AddRemote` so it fails fast but it looks like the golang http server recovers it away (but it fixes the leak at least).

**Comment by xiang90** [comment]
> the rafthttp transport stops before the peer http server stops. A new connection goes through the http server which creates a new pipeline via AddRemote. 

If the transport is stopped, we should not reply to the request and thus should not create the pipeline?

**Comment by heyitsanthony** [comment]
@xiang90 yes. Once the transport is stopped, nothing should call into it again. However, if the http server is still active, it'll still try to call into it via `rafthttp.ServeHTTP`.

**Comment by xiang90** [comment]
> However, if the http server is still active, it'll still try to call into it via rafthttp.ServeHTTP.

This is the receiver side of the rafthttp, right? Once it receives a message, it might deliver that into raft and then raft will generate output. But at that point, the sender side, which is the transporter should be stopped. No messages should be accepted I think.

File: rafthttp/transport.go [hunk 2]
```diff
@@ -202,11 +202,16 @@ func (t *Transport) Stop() {
...
 	}
+	t.peers = nil
+	t.remotes = nil
 }
 
 func (t *Transport) AddRemote(id types.ID, us []string) {
 	t.mu.Lock()
 	defer t.mu.Unlock()
+	if t.remotes == nil {
+		panic("transport stopped")
```
Comment (xiang90) [inline rafthttp/transport.go]: why not just return but panic?

**Comment by xiang90** [comment]
> A new connection goes through the http server which creates a new pipeline via AddRemote

Also it seems like etcdserver will only call addRemote during initialization? After that etcdserver should not call addRemote.

**Comment by heyitsanthony** [comment]
@xiang90 I don't understand what you're getting at. The problem is that rafthttp has no control over the http server / listener lifetime. The http handler `streamHandler` isn't bound to a connection so even if the transporter is stopped, if a new connection is established, the handler would see there is no remote and try to create a new pipeline via `AddRemote` and leak it.

rafthttp calls AddRemote because of the `X-PeerURLs` header which was a fix so a node could communicate with members that were added while it was down.

File: rafthttp/transport.go (hunk 2 above)
Comment (heyitsanthony) [inline rafthttp/transport.go]: It seems to me the transport should really only be stopped once the http server is taken down. If `AddRemote` gets called after a `Stop`, then the listener is still active and the shutdown order is being ignored.

**Comment by xiang90** [comment]
@heyitsanthony Oh. Now I understand it. I thought it was the etcdserver that add the remote back, but it is the streamHandler actually.

File: rafthttp/transport.go (hunk 2 above)
Comment (xiang90) [inline rafthttp/transport.go]: Is there a way to cleanly shutdown a http server right now? closing the listener does not mean closing all the existing connections.

Comment (heyitsanthony) [inline rafthttp/transport.go]: OK, I see how that can race if there's no clean shutdown. Will swap panic for return.

**Comment by xiang90** [comment]
lgtm

**Pull request merged**